main/pipeline/pipeline.cc

```diff
@@ -428,55 +428,107 @@ struct IndexThreadResultPack {
     IndexResult res;
 };
 
-IndexResult mergeIndexResults(const shared_ptr<core::GlobalState> cgs, const options::Options &opts,
-                              shared_ptr<BlockingBoundedQueue<IndexThreadResultPack>> input,
-                              const unique_ptr<const OwnedKeyValueStore> &kvstore) {
+struct IndexSubstitutionJob {
+    // Not necessary for substitution, but passing this through to the worker means it's freed in that thread, instead
+    // of serially in the main thread.
+    unique_ptr<core::GlobalState> threadGs;
+
+    std::optional<core::GlobalSubstitution> subst;
+    vector<ast::ParsedFile> trees;
+
+    IndexSubstitutionJob() {}
+
+    IndexSubstitutionJob(core::GlobalState &to, IndexResult res)
+        : threadGs{std::move(res.gs)}, subst{}, trees{std::move(res.trees)} {
+        if (absl::c_any_of(this->trees, [this](auto &parsed) { return !parsed.file.data(*this->threadGs).cached; })) {
+            this->subst.emplace(*this->threadGs, to);
+        } else {
+            core::GlobalSubstitution::mergeFileTables(*this->threadGs, to);
```

Comment:
I find it a little weird that we merge the file tables in two places. Should we unconditionally merge the file tables in this constructor?

Reply:
I think that's a good change -- making it so that `GlobalSubstitution` only constructs a name substitution would be a good thing.